server/src/main/java/org/elasticsearch/script/MultiVectorScoreScriptUtils.java

```diff
@@ -0,0 +1,364 @@
+                throw new IllegalArgumentException(
+                    "The query vector has an incorrect number of dimensions. Must be ["
+                        + fieldDims / 8
+                        + "] for bitwise operations, or ["
+                        + fieldDims
+                        + "] for byte wise operations: provided ["
+                        + queryVector.length
+                        + "]."
+                );
+            }
+            this.byteQueryVector = queryVector;
+            this.floatQueryVector = null;
+        }
+
+        public MaxSimBitDotProduct(ScoreScript scoreScript, MultiDenseVectorDocValuesField field, List<List<Number>> queryVector) {
+            super(scoreScript, field);
+            if (queryVector.isEmpty()) {
+                throw new IllegalArgumentException("The query vector is empty.");
+            }
+            if (field.getElementType() != DenseVectorFieldMapper.ElementType.BIT) {
+                throw new IllegalArgumentException("cannot calculate bit dot product for non-bit vectors");
+            }
+            float[][] floatQueryVector = new float[queryVector.size()][];
+            byte[][] byteQueryVector = new byte[queryVector.size()][];
+            boolean isFloat = false;
+            int lastSize = -1;
+            for (int i = 0; i < queryVector.size(); i++) {
+                if (lastSize != -1 && lastSize != queryVector.get(i).size()) {
+                    throw new IllegalArgumentException("The query vector has an inconsistent number of dimensions.");
+                }
+                lastSize = queryVector.get(i).size();
+                floatQueryVector[i] = new float[queryVector.get(i).size()];
+                if (isFloat == false) {
+                    byteQueryVector[i] = new byte[queryVector.get(i).size()];
+                }
+                for (int j = 0; j < queryVector.get(i).size(); j++) {
+                    Number number = queryVector.get(i).get(j);
+                    floatQueryVector[i][j] = number.floatValue();
+                    if (isFloat == false) {
+                        byteQueryVector[i][j] = number.byteValue();
+                    }
+                    if (isFloat
+                        || floatQueryVector[i][j] % 1.0f != 0.0f
+                        || floatQueryVector[i][j] < Byte.MIN_VALUE
+                        || floatQueryVector[i][j] > Byte.MAX_VALUE) {
+                        isFloat = true;
+                    }
+                }
+            }
+            int fieldDims = field.get().getDims();
+            if (isFloat) {
+                this.floatQueryVector = floatQueryVector;
+                this.byteQueryVector = null;
+                if (fieldDims != floatQueryVector[0].length) {
+                    throw new IllegalArgumentException(
+                        "The query vector has an incorrect number of dimensions. Must be ["
```

Comment:
Kind of like the other comments I guess

```suggestion
                        "The query vector has inner vectors whose dimensions do not match the document's inner vectors. Must be ["
```